### ISSUE TITLE
Vless fallback Nginx

### ISSUE THREAD

**Issue by hawshemi** [chika0801/Xray-examples#81]
Hello

I'm configuring a simple VLESS+TCP+TLS(Vision) with `nginx` fallback with a website.
The website files are at `/usr/share/nginx/html/`. with the nginx `sites-enabled` edited with the domain name.
SSL is obtained with `certbot`.

The config is working ok but I want to use `nginx` and fallbacks:
With the default `nginx.conf` config, the website is ok and opening (but the fallbacks don't work).
But when I copy/pasta the `nginx.conf` from the https://github.com/chika0801/Xray-examples/tree/main/VLESS-XTLS-Vision , the nginx service will return an error with:
```
Aug 18 20:36:15  server systemd[1]: Starting nginx.service - A high performance web server and a reverse proxy server...
Aug 18 20:36:15 server  nginx[4569]: 2023/08/18 20:36:15 [emerg] 4569#4569: getpwnam("nginx") failed in /etc/nginx/nginx.conf:1
Aug 18 20:36:15 server  nginx[4569]: nginx: configuration file /etc/nginx/nginx.conf test failed
Aug 18 20:36:15 server  systemd[1]: nginx.service: Control process exited, code=exited, status=1/FAILURE
Aug 18 20:36:15 server systemd[1]: nginx.service: Failed with result 'exit-code'.
Aug 18 20:36:15 server  systemd[1]: Failed to start nginx.service - A high performance web server and a reverse proxy server.
```
and `error.log`:
```
[notice] 3968#3968: using inherited sockets from "5;6;"
```

**Comment by chika0801** [comment]
You need to post your Xray and Nginx config files, to help you analyze it.

**Comment by hawshemi** [comment]
I used the exapmles `.json` files:

https://github.com/chika0801/Xray-examples/blob/main/VLESS-XTLS-Vision/config_server.json

https://github.com/chika0801/Xray-examples/blob/main/VLESS-XTLS-Vision/nginx.conf

**Comment by chika0801** [comment]
> The config is working ok but I want to use nginx and fallbacks:
> With the default nginx.conf config, the website is ok and opening (but the fallbacks don't work).
> But when I copy/pasta the nginx.conf from the https://github.com/chika0801/Xray-examples/tree/main/VLESS-XTLS-Vision , > the nginx service will return an error with:

What I see in your paragraph is that you modified the nginx files and then the nginx program ran wrong. So I asked you to post your config file.

If you just use my example and don't modify it, it works fine.

**Comment by chika0801** [comment]
> The website files are at /usr/share/nginx/html/. with the nginx sites-enabled edited with the domain name.
SSL is obtained with certbot.

What I see in your paragraph is that you're not using the reverse proxy part of my nginx configuration, you're pointing to a path on your local machine. So I don't know the reason why your nginx configuration is wrong.

**Comment by chika0801** [comment]
When you posted your nginx xray configuration, I could only do some basic checks. If you have your own needs and modified the nginx configuration, such as pointing to your VPS local directory, a basic web page placed in the directory. I can't help you if you're having problems with the page not opening or something like that. I'm not very familiar with nginx configuration files. So for your question, I emphasize the need for you to upload your config file so that others can help you. Just typing a description of the problem is not an efficient way to communicate.

**Comment by hawshemi** [comment]
After installing nginx, I replaced the `nginx.conf` file from `/etc/nginx/` to [example ](https://github.com/chika0801/Xray-examples/blob/main/VLESS-XTLS-Vision/nginx.conf)`nginx.conf` file.
Does it work without changing any other `nginx` configurations? because when I replace the file, errors with `nginx` logs come up.

What does your example `nginx.conf` do? Can I modify it in a way that fallback to nginx which has a website at the path `/usr/share/nginx/html/` ?

**Comment by hawshemi** [comment]
nginx.conf:

```
user nginx;
worker_processes auto;

error_log /var/log/nginx/error.log notice;
pid /var/run/nginx.pid;

events {
    worker_connections 1024;
}

http {
    log_format main '[$time_local] $proxy_protocol_addr "$http_referer" "$http_user_agent"';
    access_log /var/log/nginx/access.log main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ""      close;
    }

    map $proxy_protocol_addr $proxy_forwarded_elem {
        ~^[0-9.]+$        "for=$proxy_protocol_addr";
        ~^[0-9A-Fa-f:.]+$ "for=\"[$proxy_protocol_addr]\"";
        default           "for=unknown";
    }

    map $http_forwarded $proxy_add_forwarded {
        "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";
        default "$proxy_forwarded_elem";
    }

    server {
        listen              127.0.0.1:8001 proxy_protocol;
        listen              127.0.0.1:8002 http2 proxy_protocol;

        set_real_ip_from    127.0.0.1;
        real_ip_header      proxy_protocol;

# 反向代理配置 https://www.digitalocean.com/community/tools/nginx
        location / {
            sub_filter                            $proxy_host $host;
            sub_filter_once                       off;

            set $website                          MYDOMAIN.COM;
            proxy_pass                            https://$website;
            resolver                              1.1.1.1;

            proxy_set_header Host                 $proxy_host;

            proxy_http_version                    1.1;
            proxy_cache_bypass                    $http_upgrade;

            proxy_ssl_server_name                 on;

            proxy_set_header Upgrade              $http_upgrade;
            proxy_set_header Connection           $connection_upgrade;
            proxy_set_header X-Real-IP            $proxy_protocol_addr;
            proxy_set_header Forwarded            $proxy_add_forwarded;
            proxy_set_header X-Forwarded-For      $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto    $scheme;
            proxy_set_header X-Forwarded-Host     $host;
            proxy_set_header X-Forwarded-Port     $server_port;

            proxy_connect_timeout                 60s;
            proxy_send_timeout                    60s;
            proxy_read_timeout                    60s;
        }
    }
}
```

config.json:
```
{
    "log": {
        "loglevel": "warning"
    },
    "routing": {
        "domainStrategy": "IPIfNonMatch",
        "rules": [
            {
                "type": "field",
                "ip": [
                    "geoip:ir",
                    "geoip:private"
                ],
                "outboundTag": "block"
            }
        ]
    },
    "inbounds": [
        {
            "listen": "0.0.0.0",
            "port": 443,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "XXXXX",
                        "flow": "xtls-rprx-vision"
                    }
                ],
                "decryption": "none",
                "fallbacks": [
                    {
                        "dest": "8001",
                        "xver": 1
                    },
                    {
                        "alpn": "h2",
                        "dest": "8002",
                        "xver": 1
                    }
                ]
            },
            "streamSettings": {
                "network": "tcp",
                "security": "tls",
                "tlsSettings": {
                    "rejectUnknownSni": true,
                    "minVersion": "1.3",
                    "certificates": [
                        {
                            "ocspStapling": 3600,
                            "certificateFile": "/etc/letsencrypt/live/MYDOMAIN.COM/fullchain.pem",
                            "keyFile": "/etc/letsencrypt/live/MYDOMAIN.COM/privkey.pem"
                        }
                    ]
                }
            },
            "sniffing": {
                "enabled": true,
                "destOverride": [
                    "http",
                    "tls",
                    "quic"
                ]
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "freedom",
            "tag": "direct"
        },
        {
            "protocol": "blackhole",
            "tag": "block"
        }
    ],
    "policy": {
        "levels": {
            "0": {
                "handshake": 2,
                "connIdle": 120
            }
        }
    }
}

```

**Comment by chika0801** [comment]
I installed nginx on top of my VPS using the commands from the official documentation.

After installing nginx, I directly modify and replace the contents of the /etc/nginx/nginx.conf configuration file. I've tested debian 10-12 systems, and after installing the latest version of nginx as above, nginx starts fine.

You should also check whether you have changed the public and private key files of your SSL certificate in the nginx configuration to the directory where your real certificate file is located, or whether you have used the file directory and corresponding filename mentioned in my configuration.

The above are the reasons I can think of why your nginx is not working properly, you need to do your own troubleshooting and testing.

**Comment by chika0801** [comment]
> What does your example nginx.conf do? Can I modify it in a way that fallback to nginx which has a website at the path /usr/share/nginx/html/ ?

When you install nginx for the first time, it generates a static html page in its default configuration file. You can try to modify the nginx configuration yourself by referring to the contents of the configuration. I can't help you on how to change it.

**Comment by hawshemi** [comment]
> 

I know. I replace the `index.html` with my own website files with `html/css/js`. and with the default `nginx.conf` is working ok. but it breaks when I replace it with the example `nginx.conf` file.

**Comment by hawshemi** [comment]
> > What does your example nginx.conf do? Can I modify it in a way that fallback to nginx which has a website at the path /usr/share/nginx/html/ ?
> 
> When you install nginx for the first time, it generates a static html page in its default configuration file. You can try to modify the nginx configuration yourself by referring to the contents of the configuration. I can't help you on how to change it.

I know. I replace the `index.html` with my own website files with `html/css/js`. and with the default `nginx.conf` is working ok. but it breaks when I replace it with the example `nginx.conf` file.

**Comment by chika0801** [comment]
> I know. I replace the index.html with my own website files with html/css/js. and with the default nginx.conf is working ok. but it breaks when I replace it with the example nginx.conf file.

I can't help you. I'm not very familiar with nginx configuration either.

I would suggest that you don't have to put always your own html website on your VPS. In China, I personally don't think the firewall censor will manually visit your website to see what content is displayed when your website is opened.

**Comment by hawshemi** [comment]
> > I know. I replace the index.html with my own website files with html/css/js. and with the default nginx.conf is working ok. but it breaks when I replace it with the example nginx.conf file.
> 
> I can't help you. I'm not very familiar with nginx configuration either.
> 
> I would suggest that you don't have to put always your own html website on your VPS. In China, I personally don't think the firewall censor will manually visit your website to see what content is displayed when your website is opened.

Let's say we use the default nginx `index.html` file. but how about `nginx.conf` ?
when I replace it with the new one from example, and run `systemctl restart nginx`, it returns: 
```
Aug 19 10:05:49 do-uk nginx[124418]: 2023/08/19 10:05:49 [emerg] 124418#124418: getpwnam("nginx") failed in /etc/nginx/nginx.conf:1
Aug 19 10:05:49 do-uk nginx[124418]: nginx: configuration file /etc/nginx/nginx.conf test failed
```

**Comment by chika0801** [comment]
Fallback functionality for VLESS is proposed, such as using nginx to receive requests for non-VLESS protocols is. This behavior, we generally call active probing, but there are some views in China that firewalls do not actively probe. However, the developers of Xray believe that fallback to prevent active probing is necessary.

In addition to the fallback function was developed at the time, there is another use is N protocols share 443 configuration form. Only now we do not use it much and do not recommend it.

So my reason for telling you that it doesn't matter what web content you put in your nginx configuration is this.

**Comment by chika0801** [comment]
> Let's say we use the default nginx index.html file. but how about nginx.conf ?
> when I replace it with the new one from example, and run systemctl restart nginx, it returns:

You check the configuration using nginx -t, other than that thought I can't help you.

**Comment by chika0801** [comment]
Maybe you send your nginx configuration, the problems you're having, to chatgpt and ask the AI is also a way to learn how to solve problems on your own.

**Comment by hawshemi** [comment]
at the first line of `nginx.conf` file it says: `user nginx`. so I created a username nginx with `adduser nginx`.
Now there is no error and the config file is working. But the default nginx webpage wouldn't open either and in the x-ray logs it returns:
```
2023/08/19 10:04:23 [Warning] [1705116307] app/proxyman/inbound: connection ends > proxy/vless/inbound: failed to dial to 127.0.0.1:8001 > common/retry: [dial tcp 127.0.0.1:8001: connect: connection refused] > common/retry: all retry attempts failed
```

**Comment by chika0801** [comment]
> 

You need to provide your modified nginx and xray configurations so that I can analyze it for you.

**Comment by hawshemi** [comment]
After some playing around I think it's working now. now one question: when I visit my domain, it will open the default nginx webpage but it doesn't have SSL. when I manually put `https://` before the domain the nginx server shows `502 Bad Gateway` webpage. Is this normal?

config.json:
```
{
    "log": {
        "loglevel": "warning"
    },
    "routing": {
        "domainStrategy": "IPIfNonMatch",
        "rules": [
            {
                "type": "field",
                "ip": [
                    "geoip:ir",
                    "geoip:private"
                ],
                "outboundTag": "block"
            }
        ]
    },
    "inbounds": [
        {
            "listen": "0.0.0.0",
            "port": 443,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "MyId",
                        "flow": "xtls-rprx-vision"
                    }
                ],
                "decryption": "none",
                "fallbacks": [
                    {
                        "dest": "8001",
                        "xver": 1
                    },
                    {
                        "alpn": "h2",
                        "dest": "8002",
                        "xver": 1
                    }
                ]
            },
            "streamSettings": {
                "network": "tcp",
                "security": "tls",
                "tlsSettings": {
                    "rejectUnknownSni": true,
                    "minVersion": "1.3",
                    "certificates": [
                        {
                            "ocspStapling": 3600,
                            "certificateFile": "/etc/letsencrypt/live/MY.DOMAIN.COM/fullchain.pem",
                            "keyFile": "/etc/letsencrypt/live/MY.DOMAIN.COM/privkey.pem"
                        }
                    ]
                }
            },
            "sniffing": {
                "enabled": true,
                "destOverride": [
                    "http",
                    "tls",
                    "quic"
                ]
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "freedom",
            "tag": "direct"
        },
        {
            "protocol": "blackhole",
            "tag": "block"
        }
    ],
    "policy": {
        "levels": {
            "0": {
                "handshake": 2,
                "connIdle": 120
            }
        }
    }
}
```

nginx.conf:

```
user www-data;
worker_processes auto;
pid /run/nginx.pid;
error_log /var/log/nginx/error.log;
include /etc/nginx/modules-enabled/*.conf;



events {
	worker_connections 1024;
	# multi_accept on;
}

http {
    log_format main '[$time_local] $proxy_protocol_addr "$http_referer" "$http_user_agent"';
    access_log /var/log/nginx/access.log main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ""      close;
    }

    map $proxy_protocol_addr $proxy_forwarded_elem {
        ~^[0-9.]+$        "for=$proxy_protocol_addr";
        ~^[0-9A-Fa-f:.]+$ "for=\"[$proxy_protocol_addr]\"";
        default           "for=unknown";
    }

    map $http_forwarded $proxy_add_forwarded {
        "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";
        default "$proxy_forwarded_elem";
    }

    server {
        listen              127.0.0.1:8001 proxy_protocol;
        listen              127.0.0.1:8002 http2 proxy_protocol;

        set_real_ip_from    127.0.0.1;
        real_ip_header      proxy_protocol;
        location / {
            sub_filter                            $proxy_host $host;
            sub_filter_once                       off;

            set $website                          MY.DOMAIN.COM;
            proxy_pass                            https://$website;
            resolver                              1.1.1.1;

            proxy_set_header Host                 $proxy_host;

            proxy_http_version                    1.1;
            proxy_cache_bypass                    $http_upgrade;

            proxy_ssl_server_name                 on;

            proxy_set_header Upgrade              $http_upgrade;
            proxy_set_header Connection           $connection_upgrade;
            proxy_set_header X-Real-IP            $proxy_protocol_addr;
            proxy_set_header Forwarded            $proxy_add_forwarded;
            proxy_set_header X-Forwarded-For      $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto    $scheme;
            proxy_set_header X-Forwarded-Host     $host;
            proxy_set_header X-Forwarded-Port     $server_port;

            proxy_connect_timeout                 60s;
            proxy_send_timeout                    60s;
            proxy_read_timeout                    60s;
        }
    }


	sendfile on;
	tcp_nopush on;
	types_hash_max_size 2048;
	# server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;

	##
	# Gzip Settings
	##

	gzip on;



	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
 
}

```

**Comment by hawshemi** [comment]
Update:

And sometimes (1 out of 10 refreshes) on the domain, nginx returns `400 Bad Request`. nothing is in the error logs or status. any idea?

**Comment by chika0801** [comment]
> After some playing around I think it's working now. now one question: when I visit my domain, it will open the default nginx webpage but it doesn't have SSL.

I see in your nginx configuration that you use this parameter.

`include /etc/nginx/modules-enabled/*.conf;`

So you check to see if there are any other configuration files under the /etc/nginx/modules-enabled/ directory, and at this point I presume it's the configuration file under this directory that's hit that shows, for example, an nginx-based html page that you see.

I am not using this parameter in my configuration.

**Comment by chika0801** [comment]
If you want to automatically jump to https needs to add this paragraph.

```
    server {
        listen 80;
        listen [::]:80;
        return 301 https://$host$request_uri;
    }
```

> when I manually put https:// before the domain the nginx server shows 502 Bad Gateway webpage. Is this normal?


If you manually access your web page starting with https, depending on your configuration, it should be displaying the content of the site in the parameters below.

```
            set $website                          MY.DOMAIN.COM;
            proxy_pass                            https://$website;
            resolver                              1.1.1.1;
```

**Comment by chika0801** [comment]
> And sometimes (1 out of 10 refreshes) on the domain, nginx returns 400 Bad Request. nothing is in the error logs or status. any idea?

I don't know the reason for this phenomenon.

**Comment by hawshemi** [comment]
Thanks.
No there are no modules in those folders.
--

And yes, for example when I enter `set $website    www.speedtest.net` and manually enter my domain(with `https://`) it will go to speedtest.net. but when I enter my domain in `set $website` it will return 
```
400 Bad Request
Request Header Or Cookie Too Large
nginx/1.22.1
```

--
I studied the https://github.com/chika0801/Xray-examples/blob/main/VLESS-XTLS-uTLS-REALITY/steal_oneself/nginx.conf#L31. and modified my `nginx.conf` a bit to this:
```
user www-data;
worker_processes auto;
pid /run/nginx.pid;
error_log /var/log/nginx/error.log notice;
include /etc/nginx/modules-enabled/*.conf;



events {
	worker_connections 1024;
	# multi_accept on;
}

http {
    log_format main '[$time_local] $proxy_protocol_addr "$http_referer" "$http_user_agent"';
    access_log /var/log/nginx/access.log main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ""      close;
    }

    map $proxy_protocol_addr $proxy_forwarded_elem {
        ~^[0-9.]+$        "for=$proxy_protocol_addr";
        ~^[0-9A-Fa-f:.]+$ "for=\"[$proxy_protocol_addr]\"";
        default           "for=unknown";
    }

    map $http_forwarded $proxy_add_forwarded {
        "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";
        default "$proxy_forwarded_elem";
    }

    server {
        listen 80;
        listen [::]:80;
        return 301 https://$host$request_uri;
    }
    
    server {
        listen              127.0.0.1:8001 ssl proxy_protocol;
        listen              127.0.0.1:8002 ssl http2 proxy_protocol;
        ssl_certificate /etc/letsencrypt/live/MY.DOMAIN.COM/fullchain.pem;
        ssl_certificate_key /etc/letsencrypt/live/MY.DOMAIN.COM/privkey.pem;
        
        set_real_ip_from    127.0.0.1;
        real_ip_header      proxy_protocol;
        location / {
            sub_filter                            $proxy_host $host;
            sub_filter_once                       off;

            set $website                          MY.DOMAIN.COM;
            proxy_pass                            https://$website;
            resolver                              1.1.1.1;

            proxy_set_header Host                 $proxy_host;

            proxy_http_version                    1.1;
            proxy_cache_bypass                    $http_upgrade;

            proxy_ssl_server_name                 on;

            proxy_set_header Upgrade              $http_upgrade;
            proxy_set_header Connection           $connection_upgrade;
            proxy_set_header X-Real-IP            $proxy_protocol_addr;
            proxy_set_header Forwarded            $proxy_add_forwarded;
            proxy_set_header X-Forwarded-For      $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto    $scheme;
            proxy_set_header X-Forwarded-Host     $host;
            proxy_set_header X-Forwarded-Port     $server_port;

            proxy_connect_timeout                 60s;
            proxy_send_timeout                    60s;
            proxy_read_timeout                    60s;
        }
    }
    

	sendfile on;
	tcp_nopush on;
	types_hash_max_size 2048;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;


	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;

	##
	# Gzip Settings
	##

	gzip on;


	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
 
}
```
First of all, I think the auto redirect to `https` is not working as I manually enter `http://MY.DOMAIN.COM` and it shows the nginx default webpage with no SSL enabled.

Second with this configuration, when I enter `https://MY.DOMAIN.COM`, it returns `ERR_HTTP2_PROTOCOL_ERROR`.

**Comment by chika0801** [comment]
```
            set $website                          MY.DOMAIN.COM;
            proxy_pass                            https://$website;
            resolver                              1.1.1.1;
```

The URL here, is the URL you want to reverse proxy agent, not fill in your own domain name.


```
	sendfile on;
	tcp_nopush on;
	types_hash_max_size 2048;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;


	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;

	##
	# Gzip Settings
	##

	gzip on;


	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
 
```
My nginx configuration example doesn't have this part, I suggest you remove them and test again.

**Comment by hawshemi** [comment]
Thanks.
So I cannot enter my own domain instead of lovelive.jp? instead of forwarding to an external website, it shows a local website.

_(I'm from Iran, Iran's GFW is upgraded and no one knows how they are probing and blocking the IP addresses so I'm testing with all of the combinations here.)_

**Comment by chika0801** [comment]
> So I cannot enter my own domain instead of lovelive.jp? instead of forwarding to an external website, it shows a local website.

This own domain name doesn't mean a domain name that you purchased and you pointed it to the IP of your VPS and applied for an SSL certificate for that domain name.

Domain name does not mean this.

The domain name can be something like speedtest.net or some other website.

**Comment by chika0801** [comment]
> (I'm from Iran, Iran's GFW is upgraded and no one knows how they are probing and blocking the IP addresses so I'm testing with all of the combinations here.)

[VLESS-XTLS-Vision](https://github.com/chika0801/Xray-examples/tree/main/VLESS-XTLS-Vision)

This is a requirement for you to purchase a domain name and apply for an SSL certificate for that domain name.

[VLESS-XTLS-uTLS-REALITY](https://github.com/chika0801/Xray-examples/tree/main/VLESS-XTLS-uTLS-REALITY)

The main new feature of this is that you can fill in a website without buying a domain name, dest that parameter.

It has a special use, we call steal your own form, if you use this form of conditions still need you to buy a domain name in advance, for this domain name to apply for an SSL certificate.

**Comment by hawshemi** [comment]
> > So I cannot enter my own domain instead of lovelive.jp? instead of forwarding to an external website, it shows a local website.
> 
> This own domain name doesn't mean a domain name that you purchased and you pointed it to the IP of your VPS and applied for an SSL certificate for that domain name.
> 
> Domain name does not mean this.
> 
> The domain name can be something like speedtest.net or some other website.

Example:
I'm pointing my domain `123.com` to the VPS IP address (which has xray-core and Vless-Vision setup). SSL is obtained for `123.com` with letsencrypt and the config is working fine. Now I add the `nginx` to the mix (with `nginx.conf` from the [example files](https://github.com/chika0801/Xray-examples/blob/main/VLESS-XTLS-Vision/nginx.conf)) Can I set up this, when I enter `123.com` into the browser it shows the website I build and put it in `/var/www/html`, instead of forwarding it to an external website like `lovelive.jp`.

**Comment by chika0801** [comment]
They all used Vision, a flow control technique, on the confrontation check. The client uses TLS fingerprinting emulation like uTLS.

If you're using REALITY and don't want to use the steal yourself form, the question at this point is the choice of dest URLs and which ones are valid.

This all needs you to test, suggest you in Xray that ISS inside https://github.com/XTLS/Xray-core/issues/2451 help developers to complete some tests. Because the developer is not in your country, some tests still need to be done by yourselves.

Also don't forget one principle we use in China, try not to let VPS access your country's website and program. I don't know if this check is used in your country too. I see that you have geoip:ir set to block in your config file which I think is good decision. https://github.com/MasterKia/iran-hosted-domains This guide will help you.

Basically, we think that since there are many network operators in your country, each of them should use different checking methods, and you are facing a much more difficult situation than we do in China.

Technically, TCP traffic class is now dominated by Xray side, in China, we can also use UDP traffic class protocols such as Hysteria TUIC juicity, I heard that UDP traffic class protocols in your country are also blocked by some of your carriers, I don't know if it's a true news.

**Comment by chika0801** [comment]
>  into the browser it shows the website I build and put it in /var/www/html, instead of forwarding it to an external website like lovelive.jp

It is possible, but I can't change the nginx config file for you. I'm not good at this, I suggest you refer to some video tutorials and search out articles that require you to learn how to configure nginx on your own!

**Comment by chika0801** [comment]
https://github.com/chika0801/Xray-examples/issues/80#issue-1852969504

Here's a TG group with your compatriots, you might try joining it.

**Comment by hawshemi** [comment]
Yes, we are testing hard with our developers in Iran. It's madness here. MCI and TCI IPSs just block the IP address if it's a Reality server (with all security measures) after 2 days. now I'm testing Tunneling from Iran VPS to foreign VPS but with the correct fallback to a pre-built website each, so we can measure Iran's GFW probing with all combinations. 

UDP is jammed by most of the IPSs. all we have is TCP.

Thanks again for your precious help.

**Comment by chika0801** [comment]
People in Xray's TG group can't contact RPRX privately either, so I think the only way for you guys to contact RPRX is to post on github and communicate with him. It's not a very efficient way to communicate, but it's the only way right now.

If you need to ask some questions about Xray configuration parameters, which I am familiar with, feel free to post a question!
